clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -697,6 +698,8 @@ public ConsumerRecords<K, V> poll(final Duration timeout) {
                 updateAssignmentMetadataIfNeeded(timer);
                 final Fetch<K, V> fetch = pollForFetches(timer);
                 if (!fetch.isEmpty()) {
+                    sendFetches(timer);
```

Comment:
this could throw `TimeoutException` I expect (poll with zero/low timeout + addAndGet) right? before this change, not having enough time to fetch would just carry on, exit the while(timer.notExpired) and return empty records. I image we could catch the timeout that the addAndGet could throw, probably log, and just carry on (to align with the previous behaviour)? (to review on the other places where we call the sendFEtches too)

Reply:
You're right. I think this should submit the request via `add()` (vs. `addAndGet`) in the prefetch case.

Reply:
`sendFetches()` now takes a flag to determine if it should wait:

```java
    private void sendFetches(Timer timer, boolean shouldWait) {
        CreateFetchRequestsEvent event = new CreateFetchRequestsEvent(calculateDeadlineMs(timer));

        if (shouldWait) {
            applicationEventHandler.addAndGet(event);
        } else {
            applicationEventHandler.add(event);
        }
    }
```

This allows prefetch to submit the request for fetch without waiting and the blocking fetch in `pollForFetches()` to wait for it to be sent out.

Does that seem like a sufficient solution?

Thanks!